//
//  Persistence.swift
//  NewsApp
//
//  Created by Yevgenii Kryzhanivskyi on 04.08.2022.
//

import CoreData

struct PersistenceController {
    static let shared = PersistenceController()
    
    static var preview: PersistenceController = {
        let result = PersistenceController(inMemory: true)
        let viewContext = result.container.viewContext
        for _ in 0..<10 {
            let article = ArticlesDB(context: viewContext)
            article.articleTitle = "Senate backs Finland"
            article.articleDescription = "Global health officials are flagging a troubling trend among children as routine vaccination rates are the lowest theyâ€™ve been in 30 years, according to a ne..."
            article.url = "https://www.youtube.com/watch?v=9PmrVvT8-Iw"
            article.urlToImage = "https://i.ytimg.com/vi/9PmrVvT8-Iw/hqdefault.jpg"
        }
        do {
            try viewContext.save()
        } catch {
            // Replace this implementation with code to handle the error appropriately.
            // fatalError() causes the application to generate a crash log and terminate. You should not use this function in a shipping application, although it may be useful during development.
            let nsError = error as NSError
            fatalError("Unresolved error \(nsError), \(nsError.userInfo)")
        }
        return result
    }()
    
    let container: NSPersistentContainer
    
    init(inMemory: Bool = false) {
        container = NSPersistentContainer(name: "NewsApp")
        if inMemory {
            container.persistentStoreDescriptions.first!.url = URL(fileURLWithPath: "/dev/null")
        }
        container.loadPersistentStores(completionHandler: { (storeDescription, error) in
            if let error = error as NSError? {
                fatalError("Unresolved error \(error), \(error.userInfo)")
            }
        })
    }
    
    func save() {
        
        let context = container.viewContext
        
        if context.hasChanges {
            do {
                try context.save()
            } catch {
                // Show some error here
            }
        }
    }
}
